lib/buffer.js

```diff
@@ -446,7 +446,48 @@ Buffer.prototype.compare = function compare(b) {
   return binding.compare(this, b);
 };
 
-function slowIndexOf(buffer, val, byteOffset, encoding) {
+
+// Finds either the first index of `val` in `buffer` at offset >= `byteOffset`,
+// OR the last index of `val` in `buffer` at offset <= `byteOffset`
+//
+// Arguments:
+// - buffer - a Buffer to search
+// - val - a string, Buffer, or number
+// - byteOffset - an index into `buffer`; will be clamped to an int32
+// - encoding - an optional encoding, relevant is val is a string
+// - dir - true for indexOf, false for lastIndexOf
+function bidirectionalIndexOf(buffer, val, byteOffset, encoding, dir) {
+  if (typeof byteOffset === 'string') {
+    encoding = byteOffset;
+    byteOffset = undefined;
+  } else if (byteOffset > 0x7fffffff) {
+    byteOffset = 0x7fffffff;
+  } else if (byteOffset < -0x80000000) {
+    byteOffset = -0x80000000;
+  }
+  if (typeof byteOffset !== 'number' || isNaN(byteOffset)) {
```

Comment:
This shouldn't be necessary. To follow `string.lastIndexOf()` the offset should be coerced to a primitive. So for example `'abcde'.lastIndexOf('c', [1]) === -1`. Basically the `byteOffset >>= 0` below should be enough.